[SYNPY-1760] Add ability to link Grid to CurationTask (#1383)
Sage-Bionetworks/synapsePythonClient: docs/guides/extensions/curator/metadata_contribution.md

Reviewer: For christina to review

# How to Enter and Update Metadata for a Curation Task

> This guide shows how to programmatically complete file-based and record-based metadata curation tasks, including adding, editing, and validating metadata.

Reviewer 2: ```suggestion
This guide shows how to programmatically complete record-based metadata curation tasks, including adding, editing, and validating metadata.
```

Reviewer 2: After reading below, I think this guide should only focus on record based metadata.  Right now we don't have functionality to re-upload metadata to file based tasks so it would be helpful to distinguish between the two

> This guide is for **data contributors** — someone who has been assigned a `CurationTask` and needs to fill in metadata against it (see [How to Set Up Metadata Curation Workflows](metadata_curation.md)). Your job is to open the work assigned to you, make your edits, and apply them so they can be validated.

Reviewer 2: ```suggestion
This guide shows how to programmatically complete file-based and record-based metadata curation tasks, including adding, editing, and validating metadata.
```

> ## What you'll accomplish

Reviewer 2: ```suggestion
## Overview
```

By following this guide, you will:

> - List the curation tasks in your project to see what's been assigned

Reviewer 2: ```suggestion
- List curation tasks in a Synapse project
```

> - Attach to an existing Grid session for a task, or create one if none exists

Reviewer 2: I think the "Attach to an existing grid session or create one if none exists should be baked into the code and not surfaced.

Author: Could you elaborate?

> - Attach to an existing Grid session for a task, or create one if none exists

Reviewer 2: ```suggestion
```

> - Pull the grid contents down to a local CSV so you can edit in bulk

Reviewer 2: ```suggestion
- Download record-based metadata locally as a csv
```

> - Upsert your edits back into the grid (record-based tasks)

Reviewer 2: ```suggestion
- Upload the record-based metadata into Synapse
```

- Apply your changes so the administrator can validate them
- Review the validation report for your submission so you can fix issues before handing the task back (record-based tasks)

> ## Prerequisites

Reviewer 2: ```suggestion
## Requirements
```

> - Apply your changes so the administrator can validate them

Reviewer 2: ```suggestion
```

> - Review the validation report for your submission so you can fix issues before handing the task back (record-based tasks)

Reviewer 2: ```suggestion
- Generate a validation report on the updated metadata
```

> - A Synapse account with access to the project containing the curation task

Reviewer 2: ```suggestion
- A Synapse account
- Completion of the certification quiz
- A minimum of **view** access on the Synapse project
- A minimum of **edit** access on the folder containing the record-set entity. 
```

    if grid_sessions:
        latest_grid = grid_sessions[0]
        print(f"Active grid sessions: {len(grid_sessions)}. Getting latest grid session: {latest_grid.session_id}")
    else:
        print(f"No active grid session for {source_id}; creating one.")
        latest_grid = Grid(record_set_id=source_id)
        latest_grid.create()

elif isinstance(curation_task.task_properties, FileBasedMetadataTaskProperties):
    source_id = curation_task.task_properties.file_view_id
    grid_sessions = list(Grid.list(source_id=source_id))
    if grid_sessions:
        latest_grid = grid_sessions[0]
        print(f"Active grid sessions: {len(grid_sessions)}. Getting latest grid session: {latest_grid.session_id}")
    else:
        print(f"No active grid session for {source_id}; creating one.")
        latest_grid = Grid(initial_query=Query(sql=f"SELECT * FROM {source_id}"))
        latest_grid.create()
```

> **Why prefer an existing session?** Grid sessions can be owned by a team (when the administrator set `assignee_principal_id` to a team ID), so teammates can share progress on the same session instead of forking divergent copies.

> ### Step 4: Pull the grid down as a CSV

Reviewer 2: ```suggestion
### Step 4: Download record-based metadata as a local CSV
```

Download the current grid contents so you can edit them locally — in pandas, Excel, or any tool that reads CSV.

```python
csv_path = latest_grid.download_csv(destination=".", file_name="temp.csv")
print(f"Grid downloaded to: {csv_path}")
```

Open the CSV, make your edits, and save it back to a local path. For example, with pandas:

```python
import pandas as pd
import numpy as np

df = pd.read_csv(csv_path)
print(df)

> # The grid you downloaded may be empty — the columns shown here will match

Reviewer 2: ```suggestion
```

> # The grid you downloaded may be empty — the columns shown here will match

Reviewer 2: I think we can move any Q&A about schemas to a different section.

> # whatever the curation administrator configured for the task's schema.

Reviewer 2: ```suggestion

```

> # Add 4 rows of placeholder numerical values using whatever columns the grid has.

Reviewer 2: I think this particular example of editing a grid should be moved to a different place.

df = pd.DataFrame(
    np.random.randint(0, 100, size=(4, len(df.columns))),
    columns=df.columns,
)

edited_path = "./grid_edited.csv"
df.to_csv(edited_path, index=False)
```

> ### Step 5: Upsert your edits back into the grid

Reviewer 2: ```suggestion
### Step 5: Import edited record-based metadata to Synapse
```

> ### Step 5: Upsert your edits back into the grid

Reviewer 2: @andrewelamb the instructions for programmatically downloading and uploading should only be for record-sets right now.  For file-based metadata, is there a way to programmatically update the file view you are considering?

Author: I think for now we're not going to add EntityView instructions.

> `import_csv` upserts rows into the grid based on the `upsert_keys` the administrator configured on the curation task. Existing rows matching on those keys are updated; new rows are inserted.

Reviewer 2: When downloading a record-based metadata. It would be helpful to tell the user what the upsert key and required columns are.

Author: I think this would need to be it's own feature.

Reviewer 2: I think the task has the upsert key as a property?

1. A JSON schema has been bound to the RecordSet by the administrator who set up the task
2. You have entered data through a Grid session
3. The Grid session has been exported back to the RecordSet — this is the step that triggers validation and populates the RecordSet's validation_file_handle_id

If the Grid was never exported (Step 6), there is nothing to review yet.

#### Retrieve and inspect the results

Validation results live on the RecordSet itself, so you can retrieve them whether or not the Grid session is still open. Use the record_set_id from your CurationTask, re-fetch the RecordSet to pick up the latest validation_file_handle_id, and pull the detailed report as a pandas DataFrame:

```python
from synapseclient.models import RecordBasedMetadataTaskProperties, RecordSet

> if isinstance(curation_task.task_properties, RecordBasedMetadataTaskProperties):

Reviewer 3: I wonder if there should be a notice for people who set up their tasks using legacy code. When I tried out this part, my `curation_task` was set up maybe a while ago and `curation_task.task_properties` returned None

Author: I added an exception when the Synapse response doesn't have TaskProperties, and a note to the dcos here about what to do when this happens.

> - For **record-based** tasks, export the grid back to the RecordSet. This creates a new version of the RecordSet and generates the validation report that administrators can retrieve via [RecordSet.get_detailed_validation_results][synapseclient.models.RecordSet.get_detailed_validation_results].

Reviewer 2: Is the validation report only run after export_to_record_set()? 

So is there any way to get the validation report when it's still a draft?

@andrewelamb

Author: @cconrad8 Could you elaborate on this?

    ```

> - For **file-based** tasks, synchronize the grid against the file view to push annotation changes back to the underlying files.

Reviewer 2: I think file-based should be it's own section

    ```python
    latest_grid.synchronize()
    ```

> Once your changes are applied and you're done with the session, you can clean it up:

Reviewer 2: ```suggestion

```

Reviewer 2: Move the delete to a new section like an appendix.  The delete functionality isn't a step that would fall right after this (the data manager for example might want to access the same session again)

Author: I'll do this. 

Whose job should it be to delete the grid? Should it be part of cleaning up the CurationTask?
Should the contributor do it when they set the status of the Task to be completed/cancelled?

```python

> latest_grid.delete()

Reviewer 2: ```suggestion
```

Use this when you want to find tasks assigned to you, tasks in a specific state, or locate a task by name.

```python
from synapseclient.models import CurationTask

PROJECT_ID = "syn123456789"

# Find all tasks assigned to the currently logged-in user
my_tasks = list(CurationTask.list(project_id=PROJECT_ID, assigned_to_me=True))

# Find tasks assigned to specific users or teams (by principal ID)
team_tasks = list(CurationTask.list(project_id=PROJECT_ID, assignee_ids=["1234567", "7654321"]))

# Find all tasks that are currently in progress
in_progress_tasks = list(
    CurationTask.list(project_id=PROJECT_ID, state_filter=["IN_PROGRESS"])
)

# Find a task by name (list() does not support name filtering directly — filter after listing)
target_name = "AnimalMetadata_Curation"
named_tasks = [
    task
    for task in CurationTask.list(project_id=PROJECT_ID)
    if task.name == target_name
]
```

#### Option C: Fetch a task directly by ID

Use this when the administrator has given you a specific task ID.

```python
from synapseclient.models import CurationTask

curation_task = CurationTask(task_id=12345).get()
```

### Step 3: Create a Grid session for the task

> Ask the task to create a new Grid session — it picks the `record_set_id` from the task properties automatically and links the session back to the task. If the task already has an active session linked, calling this replaces the link with the new session.

Reviewer 2: The wording "Ask the task to create a new Grid session" is confusing.  

I would rephrase to "Open" or "Start" a grid session.

```python
# Use whichever task you found in Step 2, for example:

> # curation_task = all_tasks[0]               # from Option A (list all)

Reviewer 2: is this using 0 because the user should be selecting the value from the list?

Reviewer 2: I would suggest being more consistent where regardless of how the user acquires the task id, that is the input to the next step.
